### PR TITLE
GRPO packing

### PR DESCRIPTION
**Pull request by vwxyzjn**
very preliminary support for GRPO with packing

```
exp_name="base_smollm_grpo_${RANDOM}"
python open_instruct/grpo_fast.py \
    --exp_name $exp_name \
    --output_dir /weka/oe-adapt-default/costah/models/$exp_name \
    --dataset_mixer_list ai2-adapt-dev/rlvr_gsm8k_zs 1.0 \
    --dataset_mixer_list_splits train \
    --dataset_mixer_eval_list ai2-adapt-dev/rlvr_gsm8k_zs 1.0 \
    --dataset_mixer_eval_list_splits train \
    --max_token_length 256 \
    --max_prompt_token_length 256 \
    --response_length 128 \
    --pack_length 1024 \
    --number_samples_per_prompt 4 \
    --model_name_or_path HuggingFaceTB/SmolLM2-135M \
    --stop_strings "</answer>" \
    --add_r1_style_format_reward \
    --non_stop_penalty False \
    --stop_token eos \
    --penalty_reward_value 0.0 \
    --temperature 0.7 \
    --ground_truths_key ground_truth \
    --chat_template_name r1_simple_chat_postpend_think \
    --sft_messages_key messages \
    --learning_rate 3e-7 \
    --total_episodes 1000000 \
    --deepspeed_stage 2 \
    --per_device_train_batch_size 1 \
    --local_rollout_forward_batch_size 1 \
    --local_mini_batch_size 16 \
    --local_rollout_batch_size 16 \
    --num_epochs 1 \
    --actor_num_gpus_per_node 1 \
    --vllm_tensor_parallel_size 1 \
    --beta 0.01 \
    --apply_verifiable_reward true \
    --seed 3 \
    --num_evals 100 \
    --save_freq 100 \
    --reward_model_multiplier 0.0 \
    --no_try_launch_beaker_eval_jobs \
    --vllm_sync_backend gloo \
    --vllm_gpu_memory_utilization 0.5 \
    --vllm_enforce_eager \
    --gradient_checkpointing \
    --single_gpu_mode \
    --with_tracking

```

to debug.

```
for dataset_combo in \
    "gsm8k ai2-adapt-dev/rlvr_gsm8k_zs" \
; do
for learning_rate in 5e-7; do
for beta in 0.0; do
for nspp in 32; do
for m in m; do
for kl_estimator in kl3; do
for local_rollout_batch_size in 4; do
if [ $m == "half-m" ]; then
    local_mini_batch_size=$(($local_rollout_batch_size * $nspp / 2))
else
    local_mini_batch_size=$(($local_rollout_batch_size * $nspp))
fi
read -r dataset_name dataset <<< "$dataset_combo"
echo --- local_mini_batch_size=$local_mini_batch_size
echo --- num_gradient_updates=$(($local_rollout_batch_size * $nspp / $local_mini_batch_size))
exp_name="0226_packing_qwen2.5_7B_${dataset_name}_${m}_${RANDOM}"
echo $exp_name $dataset
 python open_instruct/grpo_fast.py \
    --exp_name $exp_name \
    --beta $beta \
    --local_mini_batch_size $local_mini_batch_size \
    --number_samples_per_prompt $nspp \
    --output_dir /output \
    --local_rollout_batch_size $local_rollout_batch_size \
    --kl_estimator $kl_estimator \
    --learning_rate $learning_rate \
    --dataset_mixer_list $dataset 1.0 \
    --dataset_mixer_list_splits train \
    --dataset_mixer_eval_list $dataset 16 \
    --dataset_mixer_eval_list_splits train \
    --max_token_length 4096 \
    --max_prompt_token_length 2048 \
    --response_length 2048 \
    --pack_length 4096 \
    --model_name_or_path Qwen/Qwen2.5-7B \
    --stop_strings '"</answer>"' \
    --add_r1_style_format_reward \
    --chat_template_name r1_simple_chat_postpend_think \
    --oe_eval_tasks minerva_math::hamish_zs_reasoning,bbh:cot::hamish_zs_reasoning,gsm8k::hamish_zs_reasoning,minerva_math_500::hamish_zs_reasoning,zebralogic::hamish_zs_reasoning,aime::hamish_zs_reasoning,agi_eval_english:0shot_cot::hamish_zs_reasoning,gpqa:0shot_cot::hamish_zs_reasoning \
    --non_stop_penalty False \
    --stop_token eos \
    --temperature 1.0 \
    --ground_truths_key ground_truth \
    --sft_messages_key messages \
    --total_episodes 10000000 \
    --penalty_reward_value 0.0 \
    --deepspeed_stage 2 \
    --per_device_train_batch_size 1 \
    --local_rollout_forward_batch_size 1 \
    --actor_num_gpus_per_node 4 \
    --num_epochs 1 \
    --vllm_tensor_parallel_size 1 \
    --vllm_num_engines 4 \
    --enable_prefix_caching \
    --lr_scheduler_type linear \
    --apply_verifiable_reward true \
    --seed 1 \
    --num_evals 100 \
    --save_freq 2000 \
    --reward_model_multiplier 0.0 \
    --no_try_launch_beaker_eval_jobs \
    --try_launch_beaker_eval_jobs_on_weka False \
    --gradient_checkpointing \
    --with_tracking
done
done
done
done
done
done
done

```

seems to work 

image

### CODE REVIEW

File: docs/algorithms/grpo.md [hunk 1]
```diff
@@ -5,17 +5,131 @@ GRPO is an online RL method used in [DeepSeek R1 paper](https://arxiv.org/abs/25
...
+
+The corresponding evals look good as well:
+
+![grpo_qwen2.5_7B_works_evals](static/grpo_qwen2.5_7B_works_evals.png)
+
+
+## Explanation of the logged metrics
+
+
+* `episode`: the global episode number training has gone through (e.g., `3000` means we have trained on 3000 data points already)
```
Comment (natolambert) [inline docs/algorithms/grpo.md]: 3000 prompts? 
```suggestion
* `episode`: the global episode number training has gone through (e.g., `3000` means we have trained on 3000 data points already -- in the case of RLVR that is prompts, which can repeat)
```

File: docs/algorithms/grpo.md [hunk 2]
```diff
@@ -5,17 +5,131 @@ GRPO is an online RL method used in [DeepSeek R1 paper](https://arxiv.org/abs/25
...
+* `lr`: the current learning rate
+* `epoch`: the fraction or multiple of the epoch (e.g., `2.7` means we have trained on the dataset for 2 epochs and 70% of the third epoch)
+* `objective/kl`: the KL divergence between the current policy and the reference policy (sum of the KL divergence of each response token)
+* `objective/scores`: the scores of the current response, rated by a reward model
```
Comment (natolambert) [inline docs/algorithms/grpo.md]:
```suggestion
* `objective/scores`: the scores of the current response, rated by a combination of reward model, KL penalty, and verifiable rate depending on setup
```

File: open_instruct/model_utils.py [hunk 3]
```diff
@@ -259,6 +259,34 @@ def apply_verifiable_reward(
     # return rewards and count of times we applied reward
     return rewards_tensors, (rewards_tensors > 0).sum().float().view(-1)
 
+def apply_verifiable_reward_fast(
+    decoded_responses: List[str],
+    ground_truths: List[str],
+    datasets: List[str],
+    verify_reward: int = 10,
+):
+    # compare with ground truth.
```
Comment (natolambert) [inline open_instruct/model_utils.py]:
```suggestion
    # extra util to `apply_verifiable_reward` where responses are already decoded in batch process (relative to being handled at verification time).
    # compare with ground truth.
```

File: open_instruct/rl_utils2.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    assert not any(pad_token_id in response for response in responses)
+
+    query_responses = []
+    attention_masks = []
+    response_masks = []
+    num_actions = []
+    packed_seq_lens = []
+    cur_data = []
+    cur_response_mask = []
+    cur_lengths = []
```
Comment (natolambert) [inline open_instruct/rl_utils2.py]: We create cur_lengths but never use it?

**Review by natolambert — APPROVED** [review]
Some small nits. I didn't see any issues, also did some AI code review.

File: docs/algorithms/grpo.md (hunk 2 above)
Comment (vwxyzjn) [reply docs/algorithms/grpo.md]: I changed it to `* `objective/scores`: the scores of the current response, rated by a combination of reward model and other rewards (e.g., R1 style format reward, verifiable reward, etc.)`

The metrics which include KL penalty is `objective/rlhf_reward`

**Comment by vwxyzjn** [comment]
image


Vectorized advantage calculation makes the end-to-end training throughput 2x faster. I am a bit surprised by this:

when trying to read from the same trace locally, the advantage calculation took like 1 secs, unlike the 30 secs shown in the runs. Probably some multi-threading stuff happening. Anyway this should make the end-to-end training a lot faster.
